app/src/main/java/de/xikolo/controllers/settings/SettingsFragment.kt

```diff
@@ -41,9 +49,97 @@ class SettingsFragment : PreferenceFragment() {
+                    progressDialog.setCancelable(false)
+                    progressDialog.setCanceledOnTouchOutside(false)
+                    progressDialog.setProgressStyle(android.app.ProgressDialog.STYLE_HORIZONTAL)
+                    progressDialog.max = fileCount
+                    progressDialog.show()
+
+                    val migrationCallback = object : StorageUtil.StorageMigrationCallback {
+                        override fun onProgressChanged(count: Int) {
+                            activity.runOnUiThread { progressDialog.progress = count }
+                        }
+
+                        override fun onCompleted(success: Boolean) {
+                            activity.runOnUiThread {
+                                if (success)
+                                    ToastUtil.show(R.string.dialog_storage_migration_successful)
+                                else
+                                    ToastUtil.show(R.string.error_plain)
+                                progressDialog.dismiss()
+                            }
+                        }
+                    }
+
+                    if (newStorageType == StorageUtil.StorageType.INTERNAL) {
+                        StorageUtil.migrateAsync(StorageUtil.getSdcardStorage(App.getInstance())!!,
+                            StorageUtil.getInternalStorage(App.getInstance()),
+                            migrationCallback)
+                    } else {
+                        StorageUtil.migrateAsync(StorageUtil.getInternalStorage(App.getInstance()),
+                            StorageUtil.getSdcardStorage(App.getInstance())!!,
+                            migrationCallback)
+                    }
+                }
+
+                dialog.show()
+            }
+        }
+    }
+
     override fun onCreatePreferences(bundle: Bundle?, rootKey: String?) {
         addPreferencesFromResource(R.xml.settings)
 
+        val prefs = PreferenceManager.getDefaultSharedPreferences(activity)
+
+        findPreference(getString(R.string.preference_storage)).summary = prefs.getString(getString(R.string.preference_storage), getString(R.string.settings_default_value_storage))!!
+        findPreference(getString(R.string.preference_storage)).onPreferenceChangeListener = object : Preference.OnPreferenceChangeListener {
+            override fun onPreferenceChange(p0: Preference?, p1: Any?): Boolean {
+                if (DownloadService.getInstance() != null && DownloadService.getInstance().isDownloading) {
+                    ToastUtil.show(R.string.notification_storage_locked)
+                    return false
+                }
+                return true
+            }
+        }
+
+        // Android does not support multiple external storages below KITKAT
+        // Determining the states of multiple storages requires LOLLIPOP
+        if (Build.VERSION.SDK_INT < 21 || StorageUtil.getStorages(activity).size < 2) {
```

Comment:
Only support Lollipop as minimum here. We will remove Android 4 support very soon. Then we can simply remove all these simple if statements.